### PR TITLE
api: swagger: Adjust ContainerWaitResponse error as optional

### LINKED ISSUE #43655: API doc for ContainerWaitResponse required field that is optional
**Description**

A recently introduced commit to the API docs has marked an existing field as required: the `error` field on the `ContainerWaitResponse` type https://github.com/moby/moby/commit/352d1bb8e0cc6329d3137000d6540351c0c02c46 - this is not the case, and causes issues with clients that observe the required field as mandatory to the payload.

This was detected as part of the integration tests for https://github.com/fussybeaver/bollard , which is a docker client that generates its types from moby's swagger documentation.

**Steps to reproduce the issue:**
1. Run a container: e.g. `docker run --name wait_issue -d nginx`
2. Kill the container and curl the wait command: `docker kill wait_issue && curl --unix-socket /var/run/docker.sock -X POST http://test/containers/wait_issue/wait`

**Describe the results you received:**

Received a payload with a nullable non-required field
```
{"Error":null,"StatusCode":137}
```

**Describe the results you expected:**

The API swagger docs should reflect that this is an optional field

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:           20.10.16
 API version:       1.41
 Go version:        go1.18.1
 Git commit:        aa7e414fdc
 Built:             Sat May 14 12:48:33 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server:
 Engine:
  Version:          20.10.16
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.1
  Git commit:       f756502055
  Built:            Sat May 14 12:47:27 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.6.4
  GitCommit:        212e8b6fa2f44b9c21b2798135fc6fb7c53efc16.m
 runc:
  Version:          1.1.2
  GitCommit:
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.8.2-docker)
  compose: Docker Compose (Docker Inc., 2.5.0)

Server:
 Containers: 11
  Running: 1
  Paused: 0
  Stopped: 10
 Images: 31
 Server Version: 20.10.16
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: false
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: active
  NodeID: usvdo5pr82y5qofwxpcd5gmxi
  Is Manager: true
  ClusterID: aty9vazysn5y0v0yc472gnboo
  Managers: 1
  Nodes: 1
  Default Address Pool: 10.0.0.0/8
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 192.168.1.171
  Manager Addresses:
   192.168.1.171:2377
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 212e8b6fa2f44b9c21b2798135fc6fb7c53efc16.m
 runc version:
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.15.40-1-lts
 Operating System: Arch Linux
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 15.56GiB
 Name: ragamuffin
 ID: LS6Y:UKUG:BDNN:CGV4:X6P6:M6WI:E5HM:R2J4:VG2M:B6AL:7UEZ:5Q7S
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

### PR DESCRIPTION
**Pull request by fussybeaver**
**- What I did**
Removed the `required` API swagger flag on the `ContainerWaitResponse` type `Error` property

**- How I did it**
Amended API 1.39, 1.40, 1.41 and the future swagger for 1.42 YAML description to remove the `required` flag

**- How to verify it**
Generated types for the [Bollard](https://github.com/fussybeaver/bollard) docker client and ran integration tests on the new types.

**- Description for the changelog**
Removed the `required` field in the swagger documentation for the `ContainerWaitResponse` `Error` property. 

fixes #43655 

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
hm, interesting, so I recall I added the required to make the generated type match the existing one. In this commit, the type that was moved, already had "Required: true" in the generated Go code; https://github.com/moby/moby/pull/43334/commits/e4c6ca36aea01bfebe34d09c728e60ebe35a0442

Wondering if that was a quirk in the generator that's used in this repository 🤔

**Comment by thaJeztah** [comment]
Looks indeed that the generated type differs with this patch (perhaps it's ok to change though);

```
 The result of hack/generate-swagger-api.sh differs

 diff --git a/api/types/container/wait_response.go b/api/types/container/wait_response.go
 index d2fb63aac5..84fc6afddc 100644
 --- a/api/types/container/wait_response.go
 +++ b/api/types/container/wait_response.go
 @@ -10,8 +10,7 @@ package container
  type WaitResponse struct {
  
  	// error
 -	// Required: true
 -	Error *WaitExitError `json:"Error"`
 +	Error *WaitExitError `json:"Error,omitempty"`
  
  	// Exit code of the container
  	// Required: true

 Please update api/swagger.yaml with any API changes, then
 run hack/generate-swagger-api.sh.
```

**Comment by fussybeaver** [comment]
To be clear, I see the field exists but is set to `null` in JSON on the wire. So far the Bollard client API assumes required properties to be set to a value.

There is perhaps a case to look into using the `x-nullable` vendor extension, but I haven't tested thoroughly whether this works in Bollard across the whole API.

**Comment by thaJeztah** [comment]
Yes, to be honest; at times it's a bit tricky to pick the most suitable option in swagger, as not all of them are a "perfect fit" for the go types (combinations of "nullable" fields, "omit empty", etc). I think the change shown in the diff above (at least at a glance) looks still OK, and I don't expect that change to be causing issues from our side (or at least, i'd expect the client to unmarshal both the "before" and "after" identical), so to unblock this PR, I'd suggest to;

- re-generate the type (or manually update it with the diff as shown above)
- if it's not too much trouble for you; split the changes in the `api/types/container/wait_response.go` and `api/swagger.yaml` (i.e. "current version") to a separate commit

Last one gives us a bit more flexibility to decide if we want to backport only the "documentation" changes or both the documentation and code changes to the "v20.10" release branch (which is the branch that's also used for the documentation at docs.docker.com)

**Review by ndeloof — APPROVED** [review]

**Comment by fussybeaver** [comment]
I've regenerated the types with the provided make command and split the commits. Let me know if there's anything outstanding..

**Comment by fussybeaver** [comment]
I've regenerated the types with the provided make command and split the commits. Let me know if there's anything outstanding..

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by thaJeztah** [comment]
CI failure is unrelated; I kicked the Windows CI once more, but this should be ready to merge 👍

**Comment by thaJeztah** [comment]
alright; good to go; thanks!

**Pull request merged**
